### ISSUE TITLE
NVDA fails and does not recover when when all audio devices become unavailable

### ISSUE THREAD

**Issue by cyrmax** [nvaccess/nvda#15758]
### Steps to reproduce:
* Switch NVDA to wasapi;
* When NVDA is working normaly open sound card driver settings and change sample rate;
* Or else very fastly disconnect and connect your sound card again;
* Important: this is a third-party sound card, not realtek.

### Actual behavior:
NVDA stops speaking and spams with error sounds on every key press or any speech output.

### Expected behavior:
NVDA should handle those reconnects or configuration changes correctly.

### NVDA logs, crash dumps and other attachments:
[nvda-old.log](https://github.com/nvaccess/nvda/files/13300293/nvda-old.log)

#### NVDA installed/portable/running from source:
Installed.

#### NVDA version:
NVDA alpha-29899,cb356535 (2024.1.0.29899)
Also reproduceable on any NVDA version which supports wasapi.

#### Windows version:
Windows 11 23H2 (AMD64) build 22631.2506
Also was reproduceable on Windows 10.

#### Name and version of other software in use when reproducing the issue:
Not applicable.

#### Other information about your system:
Sound card is an USB-microphone with output. Model is sE Electronics NEOM USB

#### Does the issue still occur after restarting your computer?
Yes.

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Yes. Reproduceable on any NVDA which supports wasapi.

#### If NVDA add-ons are disabled, is your problem still occurring?
Yes.

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes.

**Comment by Adriani90** [comment]
cc: @jcsteh this also seems related to WASAPI.

**Comment by jcsteh** [comment]
I assume you're changing the sample rate through some software specific to the audio device, rather than through Windows sound properties? I can't reproduce this when changing the shared sampling rate on the Advanced tab of the Windows sound properties for an audio device.

I don't really know how to action this given that I can't reproduce it. I'm not sure what your audio driver is doing in this case that is different to what the code is expecting.

**Comment by beqabeqa473** [comment]
I am also unable to reproduce a problem when changing sample_rate from windows sound properties.

Seems it is specific to some sort of soundcard.

**Comment by leonardder** [comment]
I have seen similar errors in the callback with my RD Access add-on, though I'm not sure whether that has the same error. I never bothered reporting until now because I was pretty sure it is the add-on doing something obscure. It especially happens when pausing speech from a remote session.

**Comment by antonshusharin** [comment]
Here are some more details about this issue (posting on behalf of @kirill-jjj):
Whenever the sampling rate is changed in the driver's ui, the audio device is disconnected and reconnected in around a second. During this window, NVDA attempts to use the system's default audio device for output. If it is not available, the behavior described in the first comment occures (speech goes silent and does not recover).
Interestingly, NVDA's ui sounds do switch to the original device after it is reconnected. It is only speech that is affected.

**Comment by antonshusharin** [comment]
We have also discovered that if speech is disabled with NVDA+S while the sampling rate is being changed, the issue does not occure. After speech is reenabled NVDA functions correctly.

**Comment by jcsteh** [comment]
To be clear, the external device you are tweaking is also the default audio device and there is no other default device at any point? That is, while the sample rate is switching, there is no default device at all?

**Comment by antonshusharin** [comment]
> To be clear, the external device you are tweaking is also the default audio device and there is no other default device at any point? That is, while the sample rate is switching, there is no default device at all?

Yes. When we plug in another device before changing the sample rate, NVDA correctly switches to it and then to the original device once it becomes usable again.

**Comment by cyrmax** [comment]
@antonshusharin Thank you very much for making things more clear.
When we tried to investigate the problem first time we didn't pay attention to those things.
@jcsteh @Adriani90 @beqabeqa473 Yes, we changed sample rate from a sound card driver and yes it reconnects this sound card as a new device or the same device with modified parameters.

**Comment by jcsteh** [comment]
Does this work as expected with WASAPI disabled? I'm trying to work out how we can best handle this. If it does work, does it log a bunch of errors?

**Comment by antonshusharin** [comment]
Yes, this works correctly when WASAPI is disabled. I believe there are no errors, but will need to follow up with @kirill-jjj to be sure.

**Comment by jcsteh** [comment]
Note that the errors will likely occur while the device is disconnected, so you won't hear them. You'll need to explicitly check the log.

**Comment by ultrasound1372** [comment]
Does disabling and re-enabling the audio service for a period of time cause the same thing to happen? That might provide easier steps to reproduce, and would also be more relevant on servers, which would tie into that RDP case.

**Comment by antonshusharin** [comment]
@jcsteh Here's the relevant log snippet. I was wrong about there being no errors.
[winmm_no_devices.log](https://github.com/nvaccess/nvda/files/13322045/winmm_no_devices.log)

**Comment by jcsteh** [comment]
Investigation note: The error code in the log is 0x88890001, AUDCLNT_E_NOT_INITIALIZED. I guess we fail to open the device, which is fine, but that leaves our client object in an uninitialised state. We should treat AUDCLNT_E_NOT_INITIALIZED as an indication that we should try to reopen the device.

**Comment by jcsteh** [comment]
Is NVDA configured to use a specific output device or just Microsoft Sound Mapper? If it's using Microsoft Sound Mapper, I'm puzzled as to why it isn't detecting a default device change when the device reconnects.

**Comment by kirill-jjj** [comment]
> Is NVDA configured to use a specific output device or just Microsoft Sound Mapper? If it's using Microsoft Sound Mapper, I'm puzzled as to why it isn't detecting a default device change when the device reconnects.

it doesn't always failes, the first time it failed, the second time it doesn't. I don't know what it depends on.

**Comment by jcsteh** [comment]
Understood. But to my question: Is NVDA configured to use a specific output device or just Microsoft Sound Mapper?

**Comment by kirill-jjj** [comment]
> Understood. But to my question: Is NVDA configured to use a specific output device or just Microsoft Sound Mapper?

Microsoft sound mapper.
